src/main/java/com/google/sps/storage/StorageControllerInterface.java

```diff
@@ -0,0 +1,70 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.storage;
+
+/**
+ * An interface for providing unified I/O to the storage component.
+ */
+public interface StorageControllerInterface {
+  /**
+   * Creates an entirely new database entry, with just the metadata of a commit,
+   * having the passed/failed fields empty.
+   *
+   * @param entryData a JSON in plain string format, must have a "commitHash" field.
+   * @throws IllegalArgumentException if entryData is null or doesn't have a
+   *      "commitHash" field.
+   */
+  void createNewEntry(final String entryData) throws IllegalArgumentException;
+
+  /**
+   * Updates an existing database entry, with the passed/failed information for
+   * a particular buildbot. If no entry is associated to the commit hash provided,
+   * creates a new entry from the metadata included in the {@code updateData}.
+   *
+   * @param updateData a JSON in plain string format, must have a "commitHash" field.
```

Comment:
Now this gets into the meat of the question on the decision how to represent the main types of data that the system processes (aggregated and non-aggregated build statuses). A key decision would be to agree on a common representation to share between the various components, define it, and use that appropriately in interfaces.
For example, this interface suggests that the key data of the system is represented as plain Java strings in a JSON format. This is unwieldy as it doesn't precisely define the exact JSON schema that individual components expect and may lead to a lot of unneeded transformation back-and-forth then separate modules communicate with each other.
I would suggest exploring the possibility to have a more Java-native representation of these main data objects (domain objects), such as [POJOs](https://en.wikipedia.org/wiki/Plain_old_Java_object) and using those types of objects as arguments and return types from this interface.
I'd also suggest adding the person who works on the REST API part of the project as a reviewer, so you get feedback on whether the functionality defined here is appropriate for all the use-cases that component needs to support.